packages/search/src/users/users.ts

```diff
@@ -0,0 +1,171 @@
+/* Copyright (c) 2021 Environmental Systems Research Institute, Inc.
+ * Apache-2.0 */
+
+import { getProp } from "@esri/hub-common";
+import { UserSession } from "@esri/arcgis-rest-auth";
+
+import { GraphQLClient } from "graphql-request";
+import {
+  createSessionMutation,
+  userSelfQuery,
+  userSearchQuery
+} from "./queries";
+
+import { IUser } from "@esri/arcgis-rest-types";
+import {
+  base64,
+  IDateRange,
+  SortDirection,
+  ICursorSearchResults
+} from "../types";
+
+export enum UserSortableField {
+  USERNAME = "USERNAME",
+  LAST_HUB_SESSION = "LAST_HUB_SESSION"
+}
+
+interface ISortingOption {
+  field?: UserSortableField;
+  sortDirection?: SortDirection;
+}
+
+/**
+ * Follows Edge -> Node schema yielded by GraphQL requests
+ */
+interface IEdge {
+  node: IHubUser;
+  cursor?: base64;
+}
+
+/**
+ * Interface that derives directly from the User type in Hub Profiles API
+ */
+export interface IHubUser extends IUser {
+  followedInitiatives?: any[];
+  groups?: any[];
+  teams?: any[];
+  registeredEvents?: any[];
+  lastHubSession?: number;
+  visitsLast30Days?: number;
+  visitsLast60Days?: number;
+}
+
+/**
+ * Interface that derives directly from the SearchUsersFilter input for requests to the
+ * searchUser endpoint in Hub Profiles API
+ */
+export interface ISearchUsersFilter {
+  lastHubSession?: IDateRange;
+  group?: string;
+  team?: string;
+  followedInitiative?: string;
+  registeredEvent?: string;
+}
+
+export interface IUserPagingOptions {
+  first: number;
+  after?: string;
+}
+
+export interface IUserSearchOptions {
+  pagingOptions?: IUserPagingOptions;
+  sortingOptions?: ISortingOption[];
+}
+
+/**
+ * Single service that, upon instantiation with proper authentication, exposes three endpoints
+ * for interfacing with the GraphQL API exposed by Hub Profiles API
+ */
+export class UserService {
```

Comment:
The `HubService` class is not really useful in the Hub app. It essentially duplicates the Ember service w/o the benefit of DI. Sure you could have the ember service wrap an instance of this class, but that's just means additional code to wire up in `init()`. Furthermore, b/c a user can sign in/out at any point during a session, the Ember service would either still have to pass in `authentication`  each time, or update the instance whenever a user does so. All in this class doesn't benefit Ember.

I realize that the Ember app is not the only client, and that being able to use an instance might benefit other clients, but I think some more thought needs to go into the design of that. Should there be only one instance `const hub = Hub.create(portalUrl, hubApiUrl, isPortal)`. Should it be implemented at a higher level package?

Until we hone that design I propose, at the very least, that Hub.js continue to expose the stateless fns not attached to a class.

Reply:
Trailing boolean parameters are an anti-pattern that should be avoided (https://martinfowler.com/bliki/FlagArgument.html). The service class is beneficial and allows for easy mocking and polymorphism. A little more thought just needs to be given to authentication, but it's really solved problem.

Reply:
Tom's point about wrangling the auth as users sign in / out is something we should work out sooner vs later. Here's the 20,000 ft summary, but really this is the stuff we should be discussing in a wholistic design document.

### Ember Services and Class Instances

Ember services have access to the current session from Torii, which is DI'ed in. When a function on an Ember service is called, it typically delegates to Hub.js/Rest.js function calls, and (ideally) the only logic is to pass in the current session information from Torii. This leaves the Ember service as a lean facade who's main purpose is to inject auth, and makes it very easy to mock the service from w/in Ember tests (i.e. for routes & controllers)

Adding an instance of a class (i.e. `UserService`, as written) into the Ember Service, raises the issue of when in the Ember lifecycle this get's created. If we do it in the `init()` hook, we will never have access to the session, as service initiation happens very early in the app lifecycle and we can't be assured we have a session yet. Also we'd only ever have a session if the user had previously auth'd and had token info in localStorage, or has a platform cookie.

If we do it later... well there really is no "hook" for that... so that really means doing it when a function is called... and that could be done... check if an instance exists, if not, create it. Except in Hub, we support auth'd and un-authed routes... and so if the `UserService` exposes any fn's that can be called w/o a `UserSession` then we're up against the problem of sync'ing the auth state of the app, with the auth state of the instances of these Hub.js sessions. While we could use Ember's "observer" infrastructure, but in general that's not recommended.

The alternative is to drop the `UserSession` from the constructor of the `UserService`, and expose it on the function calls... which then really calls into question the value of using a class in the first place.

There are likely some other solutions to this, but let's hash them out now, vs assuming we'll have an epiphany down the line.

Reply:
Cool. Something for us to solve and make better.

Reply:
Just to clarify, if we were to merge this PR as is we would be _introducing_ the problem that would then need solve and make better.

I've proposed a few small changes to this PR that I think I would be comfortable merging until we've had a chance to do a wholistic design to address the concerns that @ajturner, @dbouwman , and I have raised.

- drop the `UserService` and export it's functions & have them get the context they need from their arguments (something that extends `IHubRequestOptions`
- rename `searchUsers` to `searchUsersFromHub`

Then once we've figured out how "service" classes fit in to this API and how we can have top level search arguments that work for either backend (previously we've talked about using query builders that could output either GraphQL) we can add `hub.searchUsers(userQueryBuilder)` which internally do something like:

```ts
if (!this.isPoral) {
  searchUsersFromHub({ ...this.requestOptions, ...userQueryBuilder.getGraphQlParams() };
} else {
  searchUsersFromPortal({ ...this.requestOptions, ...userQueryBuilder.getPortalQueryParams() };
}
```

I'm even willing to take a stab at making the changes I've proposed above if needed, but I won't be able to get to it this week.